README.md
# Flutter Piano Pro

[![pub package](https://img.shields.io/pub/v/flutter_piano_pro.svg)](https://pub.dev/packages/flutter_piano_pro)
[![GitHub stars](https://img.shields.io/github/stars/MelihHakanPektas/flutter_piano_pro.svg?style=social)](https://github.com/MelihHakanPektas/flutter_piano_pro)
[![GitHub issues](https://img.shields.io/github/issues/MelihHakanPektas/flutter_piano_pro.svg)](https://github.com/MelihHakanPektas/flutter_piano_pro/issues)

Flutter Piano Pro is a piano widget for flutter applications. It provides a configurable keyboard interface, allowing users to interact with piano notes and receive event callbacks for their interactions.

![Flutter Piano Pro Example](https://github.com/MelihHakanPektas/flutter_piano_pro/raw/master/assets/screenshot1.png)

## Installation

Add `flutter_piano_pro` as a dependency in your pubspec.yaml file:

```yaml
dependencies:
  flutter_piano_pro: ^0.5.0
```

## Usage

Import `flutter_piano_pro.dart` and `note_names.dart` files:

```dart
import 'package:flutter_piano_pro/flutter_piano_pro.dart';
import 'package:flutter_piano_pro/note_names.dart';
```

In your widget tree, use the `PianoPro` widget to add the piano:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_piano_pro/flutter_piano_pro.dart';

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Flutter Piano Pro Demo"),
      ),
      body: Center(
        child: PianoPro(
        ),
      ),
    );
  }
}
```

## Configuration

- `noteCount` - The number of notes to display on the piano.
- `totalWidth` - The total width of the piano. If `expandHorizontal` is set to `false`, this will be the fixed width of the piano. If `expandHorizontal` is set to `true`, this is ignored.
- `expandHorizontal` - If `true`, the piano will expand horizontally to fill the available space. If `false`, the piano will have a fixed width defined by `totalWidth`.
- `buttonHeight` - The height of the piano keys.
- `showNames` - If `true`, note names will be displayed on the keys.
- `showOctaveNumber` - If `true`, octave numbers will be displayed on the keys.
- `firstNote` - The index of the first note to display. `0`= C(Do).
- `firstNoteOctave` - The octave of the first note to display.
- `noteType` - The type of notes to display (`NoteType.english`, `NoteType.romanian` etc.).
- `onTapDown` - A function that is called when a note is tapped. The `NoteModel` of the tapped note is passed as the first argument, and the pointer ID is passed as the second argument.
- `onTapUpdate` - A function that is called when a pointer that is already down moves. The `NoteModel` of the updated note is passed as the first argument, and the pointer ID is passed as the second argument.
- `onTapUp` - A function that is called when a note is released. The pointer ID is passed as the argument.

## Example Usage

Here's an example of how you might use `PianoPro` in your Flutter app:

```Dart
import 'package:flutter/material.dart';
import 'package:flutter_piano_pro/flutter_piano_pro.dart';
import 'package:flutter_piano_pro/note_model.dart';
import 'package:flutter_piano_pro/note_names.dart';

main() {runApp(const MyHomePage());}
class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});
  @override
  MyHomePageState createState() => MyHomePageState();
}
class MyHomePageState extends State<MyHomePage> {
  Map<int, NoteModel> pointerAndNote = {};
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Flutter Piano Pro Demo"),
        ),
        body: Column(
          children: [
              child: PianoPro(
                noteCount: 7, // Without black keys
                totalWidth: 500,
                expandHorizontal:
                    true, // if it's true, totalWidth = constraits.maxWidth
                buttonHeight: 250,
                showNames: true,
                showOctaveNumber: true,
                firstNote: 0, // C
                firstNoteOctave: 3,
                noteType: NoteType.english, // A, B, C, D, E, F, G
                onTapDown: (note, tapId) {
                  if (note == null) return;
                  setState(() => pointerAndNote[tapId] = note);
                  debugPrint(
                      'DOWN: note= ${note.name + note.octave.toString() + (note.isFlat ? "♭" : '')}, tapId= $tapId');
                },
                onTapUpdate: (note, tapId) {
                  if (note == null) return;
                  if (pointerAndNote[tapId] == note) return;
                  setState(() => pointerAndNote[tapId] = note);
                  debugPrint(
                      'UPDATE: note= ${note.name + note.octave.toString() + (note.isFlat ? "♭" : '')}, tapId= $tapId');
                },
                onTapUp: (tapId) {
                  setState(() => pointerAndNote.remove(tapId));
                  debugPrint('UP: tapId= $tapId');
                },
              ),
            Expanded(
              child: ListView.builder(
                itemCount: pointerAndNote.length,
                itemBuilder: (BuildContext context, int index) {
                  final keys = pointerAndNote.keys.toList();
                  return ListTile(
                    title: Text(
                      "${pointerAndNote[keys[index]]!.name}${pointerAndNote[keys[index]]!.octave}${pointerAndNote[keys[index]]!.isFlat ? "♭" : ''}",
                      textAlign: TextAlign.center,
                    ),
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

In this example, we've created a simple app that displays a piano keyboard using `PianoPro`. We've also added a `ListView` below the keyboard to display the notes that are currently being pressed.

## Contributions

Contributions are welcome! Please feel free to submit a PR or open an issue.

### Contact

If you have any questions or suggestions, feel free to contact the package maintainer, Melih Hakan Pektas, via email or through GitHub.

![Melih Hakan Pektas](https://avatars.githubusercontent.com/u/108405689?s=100&v=4)

Thank you for contributing to flutter_piano_pro!

## License

This project is licensed under the MIT License. See the [LICENSE](https://github.com/MelihHakanPektas/flutter_piano_pro/blob/main/LICENSE) file for details.


CHANGELOG.md
## 0.1.3

- initial release.

## 0.1.4

- Diffrent note names added.

## 0.2.0

- Diffrent note names added.
- Piano buttons callback added.

## 0.3.0

- Black button width added
- Example sound method changed
- Scrollbar added

## 0.4.0

- Octave number fixed
- midiNoteNumber to NoteModel added

## 0.5.0

- Scroll performance update
- scrollController field added.

## 0.6.0

- animateToScrollPosition method added


test/flutter_piano_pro_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_piano_pro/flutter_piano_pro.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/note_model.dart
import 'dart:convert';

import 'package:flutter_piano_pro/note_names.dart';

class NoteModel {
  final String name;
  final int octave;
  final int noteIndex;
  final bool isFlat;
  const NoteModel({
    required this.name,
    required this.octave,
    required this.noteIndex,
    required this.isFlat,
  });

  factory NoteModel.fromMidiNoteNumber(int midiNoteNumber, NoteType noteType) {
    final noteNames = noteType.notes;
    final octave = (midiNoteNumber ~/ 12) - 1;
    int noteIndexWithFlats = midiNoteNumber % 12;
    final flatIndexes = [1, 3, 6, 8, 10];
    final isFlat = flatIndexes.contains(noteIndexWithFlats);
    if (isFlat) ++noteIndexWithFlats;
    int noteIndex = 0;
    String name = '';
    switch (noteIndexWithFlats) {
      case 0:
        noteIndex = 0;
        break;
      case 2:
        noteIndex = 1;
        break;
      case 4:
        noteIndex = 2;
        break;
      case 5:
        noteIndex = 3;
        break;
      case 7:
        noteIndex = 4;
        break;
      case 9:
        noteIndex = 5;
        break;
      case 11:
        noteIndex = 6;
        break;
      default:
        noteIndex = 0;
        break;
    }
    if (isFlat) {
      name = "${noteNames[noteIndex - 1]}♯\n${noteNames[noteIndex]}♭";
    } else {
      name = noteNames[noteIndex];
    }
    return NoteModel(
      name: name,
      octave: octave,
      noteIndex: noteIndex,
      isFlat: isFlat,
    );
  }

  int get midiNoteNumber {
    int baseNoteNumber = 12 + octave * 12;
    if (isFlat) {
      switch (noteIndex) {
        case 1: // Db/C#
          return baseNoteNumber + 1;
        case 2: // Eb/D#
          return baseNoteNumber + 3;
        case 4: // Gb/F#
          return baseNoteNumber + 6;
        case 5: // Ab/G#
          return baseNoteNumber + 8;
        case 6: // Bb/A#
          return baseNoteNumber + 10;
      }
    } else {
      switch (noteIndex) {
        case 0: // C
          return baseNoteNumber;
        case 1: // D
          return baseNoteNumber + 2;
        case 2: // E
          return baseNoteNumber + 4;
        case 3: // F
          return baseNoteNumber + 5;
        case 4: // G
          return baseNoteNumber + 7;
        case 5: // A
          return baseNoteNumber + 9;
        case 6: // B
          return baseNoteNumber + 11;
      }
    }
    throw Exception("Invalid noteIndex or flat parameter!");
  }

  NoteModel copyWith({
    String? name,
    int? octave,
    int? noteIndex,
    bool? isFlat,
  }) {
    return NoteModel(
      name: name ?? this.name,
      octave: octave ?? this.octave,
      noteIndex: noteIndex ?? this.noteIndex,
      isFlat: isFlat ?? this.isFlat,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'name': name,
      'octave': octave,
      'noteIndex': noteIndex,
      'isFlat': isFlat,
    };
  }

  factory NoteModel.fromMap(Map<String, dynamic> map) {
    return NoteModel(
      name: map['name'] as String,
      octave: map['octave'] as int,
      noteIndex: map['noteIndex'] as int,
      isFlat: map['isFlat'] as bool,
    );
  }

  String toJson() => json.encode(toMap());

  factory NoteModel.fromJson(String source) =>
      NoteModel.fromMap(json.decode(source) as Map<String, dynamic>);

  @override
  String toString() {
    return 'NoteModel(name: $name, octave: $octave, noteIndex: $noteIndex, isFlat: $isFlat)';
  }

  @override
  bool operator ==(covariant NoteModel other) {
    if (identical(this, other)) return true;

    return other.name == name &&
        other.octave == octave &&
        other.noteIndex == noteIndex &&
        other.isFlat == isFlat;
  }

  @override
  int get hashCode {
    return name.hashCode ^
        octave.hashCode ^
        noteIndex.hashCode ^
        isFlat.hashCode;
  }

  static Map<String, int> naturalMIDINoteNumbersMap = {
    'C-2': 0,
    'D-2': 2,
    'E-2': 4,
    'F-2': 5,
    'G-2': 7,
    'A-2': 9,
    'B-2': 11,
    'C-1': 12,
    'D-1': 14,
    'E-1': 16,
    'F-1': 17,
    'G-1': 19,
    'A-1': 21,
    'B-1': 23,
    'C0': 24,
    'D0': 26,
    'E0': 28,
    'F0': 29,
    'G0': 31,
    'A0': 33,
    'B0': 35,
    'C1': 36,
    'D1': 38,
    'E1': 40,
    'F1': 41,
    'G1': 43,
    'A1': 45,
    'B1': 47,
    'C2': 48,
    'D2': 50,
    'E2': 52,
    'F2': 53,
    'G2': 55,
    'A2': 57,
    'B2': 59,
    'C3': 60,
    'D3': 62,
    'E3': 64,
    'F3': 65,
    'G3': 67,
    'A3': 69,
    'B3': 71,
    'C4': 72,
    'D4': 74,
    'E4': 76,
    'F4': 77,
    'G4': 79,
    'A4': 81,
    'B4': 83,
    'C5': 84,
    'D5': 86,
    'E5': 88,
    'F5': 89,
    'G5': 91,
    'A5': 93,
    'B5': 95,
    'C6': 96,
    'D6': 98,
    'E6': 100,
    'F6': 101,
    'G6': 103,
    'A6': 105,
    'B6': 107,
    'C7': 108
  };

  static List<int> naturalMIDINoteNumbersList = [
    0,
    2,
    4,
    5,
    7,
    9,
    11,
    12,
    14,
    16,
    17,
    19,
    21,
    23,
    24,
    26,
    28,
    29,
    31,
    33,
    35,
    36,
    38,
    40,
    41,
    43,
    45,
    47,
    48,
    50,
    52,
    53,
    55,
    57,
    59,
    60,
    62,
    64,
    65,
    67,
    69,
    71,
    72,
    74,
    76,
    77,
    79,
    81,
    83,
    84,
    86,
    88,
    89,
    91,
    93,
    95,
    96,
    98,
    100,
    101,
    103,
    105,
    107,
    108
  ];
}


lib/piano_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_piano_pro/note_model.dart';
import 'package:flutter_piano_pro/note_names.dart';

class PianoView extends StatelessWidget {
  const PianoView({
    super.key,
    required this.buttonColors,
    required this.noteType,
    required this.whiteButtonWidth,
    required this.whiteButtonHeight,
    required this.noFlatIndexes,
    required this.blackButtonWidth,
    required this.blackButtonHeight,
    required this.noteCount,
    required this.firstNote,
    required this.firstNoteOctave,
    required this.showOctaveNumber,
    required this.showNames,
  });
  final Map<int, Color>? buttonColors;
  final int noteCount;
  final int firstNote;
  final NoteType noteType;
  final double whiteButtonWidth;
  final double whiteButtonHeight;
  final List<int> noFlatIndexes;
  final double blackButtonWidth;
  final double blackButtonHeight;
  final int firstNoteOctave;
  final bool showOctaveNumber;
  final bool showNames;

  @override
  Widget build(BuildContext context) {
    var noteNames = noteType.notes;
    return Stack(
      children: [
        Row(children: [
          ...List.generate(
            noteCount,
            (i) {
              var index = i + firstNote;
              var octaveCounter = i == 0 ? 0 : (index / 7).floor();
              var currentNote = NoteModel(
                  name: noteNames[index % 7],
                  octave: firstNoteOctave + octaveCounter,
                  noteIndex: index % 7,
                  isFlat: false);
              return Container(
                margin:
                    EdgeInsets.symmetric(horizontal: whiteButtonWidth / 100),
                padding: const EdgeInsets.symmetric(horizontal: 1, vertical: 8),
                width: whiteButtonWidth - ((whiteButtonWidth / 100) * 2),
                height: whiteButtonHeight,
                decoration: BoxDecoration(
                    color: buttonColors?[currentNote.midiNoteNumber] ??
                        Colors.grey,
                    borderRadius: BorderRadius.only(
                        bottomRight: Radius.circular(whiteButtonWidth / 7),
                        bottomLeft: Radius.circular(whiteButtonWidth / 7))),
                child: Align(
                  alignment: Alignment.bottomCenter,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      showOctaveNumber
                          ? Text(
                              currentNote.octave.toString(),
                              softWrap: false,
                              overflow: TextOverflow.visible,
                              style: TextStyle(
                                  color: Colors.black,
                                  fontWeight: FontWeight.bold,
                                  fontSize: whiteButtonWidth / 3 > 50
                                      ? 50 / 2
                                      : (whiteButtonWidth / 3) / 2),
                            )
                          : const SizedBox(),
                      showNames
                          ? Text(
                              currentNote.name,
                              softWrap: false,
                              overflow: TextOverflow.visible,
                              style: TextStyle(
                                  color: Colors.black,
                                  fontWeight: FontWeight.bold,
                                  fontSize: whiteButtonWidth / 3 > 45
                                      ? 45
                                      : whiteButtonWidth / 3),
                            )
                          : const SizedBox(),
                    ],
                  ),
                ),
              );
            },
          )
        ]),
        Row(
          children: [
            ...List.generate(
              noteCount,
              (i) {
                var index = i + firstNote;
                var octaveCounter = i == 0 ? 0 : (index / 7).floor();
                if (index % 7 == 0 && i != 0) ++octaveCounter;
                if (!noFlatIndexes.contains((index % 7)) && i != 0) {
                  var currentNote = NoteModel(
                    name:
                        "${noteNames[(index % 7) - 1]}♯\n${noteNames[(index % 7)]}♭",
                    octave: firstNoteOctave + octaveCounter,
                    noteIndex: index % 7,
                    isFlat: true,
                  );
                  return Row(
                    children: [
                      Container(
                        margin: EdgeInsets.symmetric(
                            horizontal: whiteButtonWidth / 100),
                        padding: const EdgeInsets.symmetric(vertical: 8),
                        width: blackButtonWidth - (whiteButtonWidth / 100) * 2,
                        height: blackButtonHeight,
                        decoration: BoxDecoration(
                            color: buttonColors?[currentNote.midiNoteNumber] ??
                                Colors.black,
                            borderRadius: BorderRadius.only(
                                bottomLeft:
                                    Radius.circular(blackButtonWidth / 7),
                                bottomRight:
                                    Radius.circular(blackButtonWidth / 7))),
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            showOctaveNumber
                                ? Text(
                                    currentNote.octave.toString(),
                                    textAlign: TextAlign.center,
                                    softWrap: false,
                                    overflow: TextOverflow.visible,
                                    style: TextStyle(
                                        color: Colors.white,
                                        fontWeight: FontWeight.bold,
                                        fontSize: blackButtonWidth / 4 > 16
                                            ? 16
                                            : blackButtonWidth / 4),
                                  )
                                : const SizedBox(),
                            showNames
                                ? Flexible(
                                    child: Text(
                                      currentNote.name,
                                      textAlign: TextAlign.center,
                                      softWrap: false,
                                      overflow: TextOverflow.visible,
                                      style: TextStyle(
                                          color: Colors.white,
                                          fontWeight: FontWeight.bold,
                                          fontSize: blackButtonWidth / 4 > 16
                                              ? 16
                                              : blackButtonWidth / 4),
                                    ),
                                  )
                                : const SizedBox(),
                          ],
                        ),
                      ),
                      !(i == noteCount - 1)
                          ? SizedBox(
                              width: whiteButtonWidth - (blackButtonWidth),
                            )
                          : const SizedBox(),
                    ],
                  );
                } else {
                  return i == 0
                      ? SizedBox(
                          width: whiteButtonWidth - (blackButtonWidth / 2),
                        )
                      : SizedBox(
                          width: whiteButtonWidth,
                        );
                }
              },
            )
          ],
        )
      ],
    );
  }
}


lib/piano_scrollbar.dart
import 'package:flutter/material.dart';

class PianoScrollbar extends StatefulWidget {
  const PianoScrollbar({
    Key? key,
    required this.scrollController,
    required this.scrollWidth,
    required this.constraints,
    required this.scale,
    required this.wButtonH,
  }) : super(key: key);

  final ScrollController scrollController;
  final double scrollWidth;
  final Size constraints;
  final double scale;
  final double wButtonH;

  @override
  State<PianoScrollbar> createState() => PianoScrollbarState();
}

class PianoScrollbarState extends State<PianoScrollbar> {
  double scrollPosition = 0.0;

  void jumpToScrollPosition(double position) {
    var position0 = position;
    if (position < 0) position0 = 0;
    if (position > widget.constraints.width - widget.scrollWidth) {
      position0 = widget.constraints.width - widget.scrollWidth;
    }
    setState(() {
      scrollPosition = position0;
    });
    widget.scrollController.jumpTo(position0 * widget.scale);
  }

  void animateToScrollPosition(double position) {
    var position0 = position;
    if (position < 0) position0 = 0;
    if (position > widget.constraints.width - widget.scrollWidth) {
      position0 = widget.constraints.width - widget.scrollWidth;
    }
    setState(() {
      scrollPosition = position0;
    });
    widget.scrollController.animateTo(position0 * widget.scale,
        duration: const Duration(milliseconds: 200), curve: Curves.ease);
  }

  void handleDragUpdate(DragUpdateDetails details) {
    final dx = details.delta.dx;
    final newPosition = scrollPosition + dx;

    if (newPosition <= 0 && dx <= 0) return;

    if (newPosition + widget.scrollWidth >= widget.constraints.width &&
        dx >= 0) {
      jumpToScrollPosition(scrollPosition);
      return;
    }

    if (newPosition < 0) {
      jumpToScrollPosition(0);
      return;
    }

    if (newPosition + widget.scrollWidth > widget.constraints.width) {
      jumpToScrollPosition(widget.constraints.width - widget.scrollWidth);
      return;
    }

    jumpToScrollPosition(newPosition);
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onHorizontalDragDown: (details) {
        final position = details.localPosition.dx - (widget.scrollWidth / 2);
        if (position <= 0) {
          animateToScrollPosition(0);
        } else if (position + widget.scrollWidth >= widget.constraints.width) {
          animateToScrollPosition(
              widget.constraints.width - widget.scrollWidth);
        } else {
          animateToScrollPosition(position);
        }
      },
      onHorizontalDragUpdate: handleDragUpdate,
      child: Row(
        mainAxisSize: MainAxisSize.max,
        children: [
          Container(
            color: Colors.black.withOpacity(0.5),
            width: scrollPosition,
            height: widget.wButtonH,
          ),
          Container(
            height: widget.wButtonH,
            width: widget.scrollWidth,
            color: Colors.transparent,
          ),
          Container(
            color: Colors.black.withOpacity(0.5),
            width: widget.constraints.width -
                (scrollPosition + widget.scrollWidth),
            height: widget.wButtonH,
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_midi_pro/flutter_midi_pro.dart';
import 'package:flutter_piano_pro/flutter_piano_pro.dart';
import 'package:flutter_piano_pro/note_model.dart';
import 'package:flutter_piano_pro/note_names.dart';
import 'package:flutter_piano_pro/piano_scrollbar.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatefulWidget {
  const MainApp({super.key});

  @override
  State<MainApp> createState() => _MainAppState();
}

class _MainAppState extends State<MainApp> {
  double noteCount = 7;
  double width = 300;
  double buttonWidthRatio = 2;
  double height = 250;
  bool expand = true;
  int firstNote = 0;
  int firstNoteOctave = 3;
  NoteType noteType = NoteType.english;
  bool showNames = true;
  bool showOctaveNumbers = true;
  Map<int, NoteModel> pointerAndNote = {};
  String sf2Path = 'assets/tight_piano.sf2';
  final _midi = MidiPro();
  final scrollBarKey = GlobalKey<PianoScrollbarState>();

  void animateToScrollPosition(double position) {
    scrollBarKey.currentState?.animateToScrollPosition(position);
  }

  void play(int midi, {int velocity = 127}) {
    _midi.playMidiNote(midi: midi, velocity: velocity);
  }

  void stop(int midi) {
    _midi.stopMidiNote(midi: midi);
  }

  @override
  void initState() {
    _midi.loadSoundfont(sf2Path: sf2Path);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    List<String> noteNames = noteType.notes;
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(colorSchemeSeed: Colors.brown),
      home: Scaffold(
        appBar: AppBar(actions: [
          IconButton(
              onPressed: () => animateToScrollPosition(400),
              icon: Icon(Icons.abc))
        ], title: const Text('Flutter Piano Pro')),
        body: ListView(
          physics: const NeverScrollableScrollPhysics(),
          children: [
            ListTile(
              title: const Text('Note Names'),
              trailing: SizedBox(
                width: 250,
                child: DropdownButton(
                    value: noteType,
                    items: List.generate(NoteType.values.length, (index) {
                      return DropdownMenuItem(
                          value: NoteType.values[index],
                          child: Text(NoteType.values[index].notes.toString()));
                    }),
                    onChanged: (value) => setState(() => noteType = value!)),
              ),
            ),
            const Divider(
              height: 0,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    const Text('First Note'),
                    SizedBox(
                      child: DropdownButton(
                          value: firstNote,
                          items: List.generate(noteNames.length, (index) {
                            return DropdownMenuItem(
                                value: index,
                                child: Text(noteNames[index].toString()));
                          }),
                          onChanged: (value) =>
                              setState(() => firstNote = value!)),
                    ),
                  ],
                ),
                Column(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    const Text('First Octave'),
                    SizedBox(
                      child: DropdownButton(
                          value: firstNoteOctave,
                          items: List.generate(noteNames.length, (index) {
                            return DropdownMenuItem(
                                value: index, child: Text(index.toString()));
                          }),
                          onChanged: (value) =>
                              setState(() => firstNoteOctave = value!)),
                    ),
                  ],
                ),
              ],
            ),
            const Divider(
              height: 0,
            ),
            ListTile(
              leading: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  const Text(
                    'Note Count',
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  Text(noteCount.round().toInt().toString()),
                ],
              ),
              trailing: SizedBox(
                width: 250,
                child: Slider(
                  divisions: 44,
                  max: 45,
                  min: 1,
                  value: noteCount,
                  onChanged: (value) => setState(() => noteCount = value),
                ),
              ),
            ),
            const Divider(
              height: 0,
            ),
            ListTile(
              leading: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  const Text(
                    'Height',
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  Text(height.round().toInt().toString()),
                ],
              ),
              trailing: SizedBox(
                width: 250,
                child: Slider(
                  divisions: 27,
                  max: 300,
                  min: 150,
                  value: height,
                  onChanged: (value) => setState(() => height = value),
                ),
              ),
            ),
            const Divider(
              height: 0,
            ),
            ListTile(
              leading: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  const Text(
                    'Black Button\nWidth Ratio',
                    textAlign: TextAlign.center,
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  Text(buttonWidthRatio.toStringAsFixed(2)),
                ],
              ),
              trailing: SizedBox(
                width: 250,
                child: Slider(
                  divisions: 40,
                  max: 3,
                  min: 1,
                  value: buttonWidthRatio,
                  onChanged: (value) =>
                      setState(() => buttonWidthRatio = value),
                ),
              ),
            ),
            const Divider(
              height: 0,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    const Text(
                      'Width',
                      style: TextStyle(fontWeight: FontWeight.bold),
                    ),
                    Text(width.round().toInt().toString()),
                  ],
                ),
                SizedBox(
                  width: 200,
                  child: Slider(
                    divisions: 50,
                    max: 5000,
                    min: 100,
                    value: width,
                    onChanged: expand
                        ? null
                        : (value) => setState(() => width = value),
                  ),
                ),
                Column(
                  children: [
                    const Text('Expand Vertical'),
                    Checkbox(
                        value: expand,
                        onChanged: (value) => setState(() => expand = value!)),
                  ],
                )
              ],
            ),
            const Divider(
              height: 0,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    const Text('Names'),
                    Checkbox(
                        value: showNames,
                        onChanged: (value) =>
                            setState(() => showNames = value!)),
                  ],
                ),
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    const Text('Octave'),
                    Checkbox(
                        value: showOctaveNumbers,
                        onChanged: (value) =>
                            setState(() => showOctaveNumbers = value!)),
                  ],
                ),
              ],
            ),
            PianoPro(
                scrollBarKey: scrollBarKey,
                onTapDown: (NoteModel? note, int tapId) {
                  if (note == null) return;
                  play(note.midiNoteNumber);
                  pointerAndNote[tapId] = note;
                },
                onTapUpdate: (NoteModel? note, int tapId) {
                  if (note == null) return;
                  if (pointerAndNote[tapId] == note) return;
                  stop(pointerAndNote[tapId]!.midiNoteNumber);
                  play(note.midiNoteNumber);
                  pointerAndNote[tapId] = note;
                },
                onTapUp: (int tapId) {
                  stop(pointerAndNote[tapId]!.midiNoteNumber);
                  pointerAndNote.remove(tapId);
                },
                noteType: noteType,
                showNames: showNames,
                showOctave: showOctaveNumbers,
                expand: expand,
                whiteWidth: width,
                firstNoteIndex: firstNote,
                firstOctave: firstNoteOctave,
                whiteHeight: height,
                blackWidthRatio: buttonWidthRatio,
                scrollController: ScrollController(),
                noteCount: noteCount.round().toInt()),
          ],
        ),
      ),
    );
  }
}
